File: CSX/Closure, Scope & Execution Context/after.js
// AFTER
// expected input: number, callbackFn
// expected ouptut: evaluated result of callbackFn(...args)

// approach: store counter in outer function. if counter is met, return inner function.

function after(num, cb) {
  let counter = 0;

  function afteredCb(...args) {
    counter++;
    if (counter >= num) return cb(...args);
  }

  return afteredCb;
}

const called = function(string) { return('hello ' + string); };
const afterCalled = after(3, called);

console.log(afterCalled('world'));
console.log(afterCalled('world'));
console.log(afterCalled('world'));

File: CSX - Python/Closure, Scope & Execution Context/after.py
def after(n, cb):
    counter = 0
    def inner(*args):
        nonlocal counter
        counter += 1
        return cb(*args) if counter >= n else None
    return inner

def called(s):
    return 'hello ' + s

after_called = after(3, called)

print(after_called('world'))
print(after_called('world'))
print(after_called('world'))
